### ISSUE TITLE
 RDS Proxy Support

### ISSUE THREAD

**Issue by gazoakley** [terraform-providers/terraform-provider-aws#12690]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon RDS Proxy is a fully managed, highly available database proxy for Amazon Relational Database Service (RDS). RDS Proxy makes applications more scalable, more resilient to database failures, and more secure.

### New or Affected Resource(s)

* `aws_db_proxy`
* `aws_db_proxy_default_target_group`
* `aws_db_proxy_target`

### Potential Terraform Configuration

```hcl
resource "aws_db_proxy" "example" {
  auth {
    auth_scheme = "SECRETS"
    description = "example"
    iam_auth    = "DISABLED"
    secret_arn  = ""
    username    = "example"
  }

  db_proxy_name       = "example"
  debug_logging       = false
  engine_family       = "MYSQL"
  idle_client_timeout = 60
  require_tls         = true
  role_arn            = "aws:"

  tags = {
    Name = "example"
    Key  = "value"
  }

  vpc_security_group_ids = ["sg-12345678901234567"]
  vpc_subnet_ids         = ["subnet-12345678901234567"]
}


resource "aws_db_proxy_default_target_group" "example" {
  connection_pool_config {
    connection_borrow_timeout    = 120
    init_query                   = "SET x=1, y=2"
    max_connections_percent      = 100
    max_idle_connections_percent = 50
    session_pinning_filters      = ["SELECT"]
  }

  db_proxy_name     = aws_db_proxy.example.db_proxy_name
}

resource "aws_db_instance" "example" {
  # ...
}

resource "aws_db_proxy_target" "example" {
  # db_cluster_identifier  = ""
  db_instance_identifier = aws_db_instance.example.name
  db_proxy_name          = aws_db_proxy.example.db_proxy_name
  target_group_name      = aws_db_proxy_default_target_group.example.name
}
```

### References

* https://aws.amazon.com/about-aws/whats-new/2019/12/amazon-rds-proxy-available-in-preview/
* https://aws.amazon.com/about-aws/whats-new/2020/04/amazon-rds-proxy-with-postgresql-compatibility-preview/

**Comment by phanimullapudi** [comment]
Hi - We are currently blocked using RDS proxy @ Adobe due to resource not available in TF. All our AWS code is TF based and we were planning to use RDS Proxy + Aurora.

**Comment by gazoakley** [comment]
@phanimullapudi Please add a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) here: https://github.com/terraform-providers/terraform-provider-aws/issues/12690#issue-595077362

**Comment by phanimullapudi** [comment]
👍

**Comment by jrobison-sb** [comment]
This is now generally available, no longer in preview.

https://aws.amazon.com/blogs/aws/amazon-rds-proxy-now-generally-available/

**Comment by brannondorsey** [comment]
Any updates on when support for this could be expected? Like @phanimullapudi some of our work is blocked until the AWS TF provider supports RDS proxy. Thanks!

**Comment by jrobison-sb** [comment]
@apparentlymart @bflad Is there any reason this couldn't merge? It seems like the tests pass. We're also in a position where we would use this service if we could, but it's blocked until Terraform supports it.

**Comment by bflad** [comment]
Hi folks 👋  We are just as excited as you are that this functionality is now generally available. We are beginning work on [breaking changes as part of the version 3.0.0 release of the Terraform AWS Provider](https://github.com/terraform-providers/terraform-provider-aws/milestone/70) now. The earliest consideration from the HashiCorp maintainers for review of existing contributions would be after 3.0 is released in (hopefully) two or three weeks. Any prioritization over [existing roadmap priorities](https://github.com/terraform-providers/terraform-provider-aws/blob/master/ROADMAP.md) would need to be made based on our criteria which can be found in the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) and we are actively working on next quarter's roadmap so getting those 👍  upvotes in is very helpful.

**Comment by alexjurkiewicz** [comment]
Please break out the `auth` block to a separate resource, like:
```hcl
resource aws_db_proxy_authorisation myuser {
  proxy_identifier = aws_db_proxy.default.id
  secret_arn = aws_secretsmanager_secret.myuser.arn
}
```

I see this situation as analogous to the way security group rules were broken out from the security group resource.

If secrets are managed inline as part of the `aws_db_proxy` resource, it will be impossible to dynamically manage the set of users independently of the Proxy resource itself. Here's an example workflow that would make use of this:

1. Stack A creates RDS database cluster & proxy resource
2. Stack B create a new DB user in the RDS cluster
3. To use this user with the proxy, you need to hardcode the username into stack A. Circular dependency!

**Comment by gazoakley** [comment]
Hi @alexjurkiewicz - I don't think the AWS API deals well with this unfortunately:

* An array of `UserAuthConfig` objects must be provided when calling `CreateDBProxy` - I don't know if that could be an empty array, but I suspect it won't work
* There's no API to add/remove the authentication settings - instead there's a `ModifyDBProxy` call that replaces the settings as a whole. A resource designed to work as above would have to pull the existing settings, update them in memory and push them back - but that would likely result in race conditions (two `aws_db_proxy_authorisation` resources trying to call `ModifyDBProxy` on the same proxy would suffer from last write wins)

@bflad - Any design suggestions here?

**Comment by bflad** [comment]
Hi @gazoakley 👋  This functionality is on the [maintainer's team roadmap for this quarter](https://github.com/terraform-providers/terraform-provider-aws/blob/master/ROADMAP.md), but we haven't set aside a specific timeline to look into it yet. I think we might be able to look into this in about two weeks given some other in-flight work happening.

Generally speaking though, if the create API requires certain configuration (as it appears it does in this case), that configuration and handling has to live with the "parent" resource. Attempting to create a "child" association resource in this situation leads to complicated implementation details like we have with the [`aws_route53_zone_association` resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_zone_association). We also have this problem with ECS API in places as well. It is not something that we can workaround due to the API and Terraform Core/SDK design requirements.

**Comment by alexjurkiewicz** [comment]
I agree this design results in a more complex API for users ("do I use the X block or resource?"), but it supports a broader range of use cases. I see RDS Proxy as being used like security groups, where a "container" proxy is created and output in one Terraform configuration, then authorisations are added to it from other configurations. I hope you decide to support this 🙏

**Comment by bflad** [comment]
The first resource, `aws_db_proxy`, has been merged and will release with version 3.5.0 of the Terraform AWS Provider, likely on Thursday. Thanks to @gazoakley for the implementation! 👍  We'll be working through other potential resources the next few days/weeks.

**Comment by bflad** [comment]
The second (optional) resource, `aws_db_proxy_default_target_group`, was released with version 3.6.0 of the Terraform AWS Provider last week. Going to provide a review for `aws_db_proxy_target` today.

**Comment by nirsv** [comment]
How can I use it as data? I see that data "aws_db_proxy" is not supported. Is there any workaround?

**Comment by bflad** [comment]
Hi @nirsv 👋  If you have a use case for looking up RDS DB Proxies via a data source, it might be best to capture that (and other future) functionality in a new, separate feature request. This one will be closed once the `aws_db_proxy_target` resource is merged to round out the initial end-to-end configuration of this feature and prevent this generic issue from not having a definition of "done". 👍

**Comment by bflad** [comment]
Support for the `aws_db_proxy_target` resource has been merged and will release with version 3.9.0 of the Terraform AWS Provider, likely tomorrow. Thanks to @goakley for the implementation. 👍 

If you have bug reports or additional feature requests with the RDS DB Proxy support, please submit new GitHub issues following the templates and we will take a look. 😄
